### ISSUE TITLE
Improve Pulumi update performance

### ISSUE THREAD

**Issue by chrsmith** [pulumi/pulumi#2674]
When running `pulumi update`, sometimes the overall operation can be "slow". There are several root causes for this, such as the specifics of the protocol used to perform updates, request latency combined with the number of requests, and specifics of how the resource providers work.

This is the tracking issue for the set of performance improvements we want to make in the CLI, which will probably be broken down into concrete work items after we have the profiling data to figure out where to best spend our time.

**Comment by chrsmith** [comment]
For milestone 0.23 we made some minor changes in the `pulumi` CLI. Server-side we made some changes to cut down on response latency, as well as add more logging for long-tail requests.

There isn't much "low hanging fruit" left to improve performance without changing the stack update protocol. Moving to the next milestone to continue making what progress we can, and lay any groundwork for when we do rev that protocol.

**Comment by chrsmith** [comment]
Landed https://github.com/pulumi/pulumi/pull/2858, which should greatly reduce network resource usage, and also reduce the impact of latency when doing updates outside of the US.

**Comment by lukehoban** [comment]
We did a [deep dive investigation](https://docs.google.com/document/d/1CCpCipGuTlZKC-znpm_YAmutXja0GiTpEJ1lUgvdVdI/edit?usp=sharing) to understand opportunities to improve performance.

Copying from the intro to that doc:

> This doc includes a deep dive on end-to-end performance of a single “worst case but still real world” update.

> This example update was a deployment of an app which deploys a GKE cluster along with KNative, resulting in ~300 resource being constructed total - mostly by volume Kuberentes resources.  The deployment was done from a $5 Digital Ocean Droplet in India, deploying to Google’s us-central1-a zone, and saving checkpoints to the app.pulumi.com backend (in Oregon).  The deployment completed successfully in 800s (~13 minutes)

> TL;DR It looks like for this update, 90% of the wall clock time was accounted for by cloud resource provisioning (and cannot be improved at the CLI/Service level).  Optimization of CLI/Service interaction could potentially shave 5-10% off total deployment time (40-80s).  

The work in #2858 should contribute to saving ~29s, a large portion of the time we believe is possible to save here.

We plan to re-run this analysis in low-bandwidth conditions as well, to understand whether that can cause significantly different results here.  But for now, we feel good that there is no smoking gun on E2E update performance, and we have addressed the one problem that we did have concrete evidence on.
